src/anchor-navigation/scroll-spy.tsx

```diff
@@ -0,0 +1,95 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+import { useCallback, useEffect, useState } from 'react';
+
+const isBrowser = typeof window !== 'undefined';
+
+export default function useScrollSpy({
+  hrefs,
+}: {
+  hrefs: string[];
+}): [
+  string | undefined,
+  React.Dispatch<React.SetStateAction<string | undefined>>,
+  React.Dispatch<React.SetStateAction<boolean>>
+] {
+  const [currentHref, setCurrentHref] = useState<string>();
+  const [scrollY, setScrollY] = useState(isBrowser ? window.pageYOffset : 0);
+  const [disableTracking, setDisableTracking] = useState(false);
+
+  // Value, in pixels, accounting for some padding in the scroll spy logic
+  const EXTRA_OFFSET = 200;
+
+  // Scroll event handler
+  const updateScroll = useCallback(() => {
+    if (isBrowser) {
+      setScrollY(window.pageYOffset);
+    }
+  }, []);
+
+  // Get the bounding rectangle of an element by href
+  const getRectByHref = useCallback(href => {
+    const element = document.getElementById(href.slice(1));
+    return element?.getBoundingClientRect();
+  }, []);
+
+  // Check if we're scrolled to the bottom of the page
+  const isPageBottom = useCallback(() => {
+    const lastAnchorId = hrefs[hrefs.length - 1]?.slice(1);
+    const lastAnchorElementExists = document.getElementById(lastAnchorId);
+    return !!lastAnchorElementExists && scrollY >= Math.floor(document.body.scrollHeight - window.innerHeight);
+  }, [scrollY, hrefs]);
+
+  // Find the href for which the element is within the viewport plus EXTRA_OFFSET
+  const findHrefInView = useCallback(() => {
+    return hrefs.find(href => {
+      const rect = getRectByHref(href);
+      return rect && rect.bottom <= window.innerHeight && rect.bottom - EXTRA_OFFSET >= 0;
```

Comment:
I've noticed that when clicking on an anchor for which the content is shorter than the viewport, the next section will be activated instead (you can try this on the examples you provide in this PR using a big monitor: clicking e.g, on section 1.1 will highlight section 1.1.1).

Note that `rect.bottom` returns the coordinate _from the top of the viewport_ (not from the bottom) of the element's bottom edge, therefore `rect.bottom - EXTRA_OFFSET` will be negative for the element that we scrolled to the top. This might be leading to this function not returning any element and then falling back to `findLastHrefInView` as per the logic in line 86 below.

Reply:
I've noticed testing. `EXTRA_OFFSET` is now the `scrollSpyOffset`, which should fix this.
I'll make sure to add integration tests covering these edge cases.